library/src/permutation/permutation_solution_impl.hpp

```diff
@@ -0,0 +1,190 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ *******************************************************************************/
+
+#ifndef HIPTENSOR_PERMUTATION_SOLUTION_IMPL_HPP
+#define HIPTENSOR_PERMUTATION_SOLUTION_IMPL_HPP
+
+#include <numeric>
+#include <map>
+
+#include "permutation_solution.hpp"
+#include "hash.hpp"
+
+namespace std
+{
+    template <>
+    struct hash<hiptensor::PermutationSolution>
+    {
+        size_t operator()(hiptensor::PermutationSolution const& s) const noexcept
+        {
+            return hash<hiptensor::PermutationSolutionParams>{}(*s.params(), s.threadDim());
+        }
+    };
+}
+
+namespace hiptensor
+{
+    template <typename DeviceOp, typename Enabler = void>
+    class PermutationSolutionImpl;
+
+    template <typename DeviceOp>
+    class PermutationSolutionImpl<DeviceOp>
+        : public PermutationSolution
+    {
+    public:
+        PermutationSolutionImpl(std::unique_ptr<DeviceOp>&& deviceOp)
+            : PermutationSolution(std::move(deviceOp),
+                                  std::make_unique<PermutationSolutionParamsImpl<DeviceOp>>())
+        {
+        }
+
+        bool initArgs(void const*                       alpha,
+                      void const*                       A,
+                      void*                             B,
+                      std::vector<std::size_t> const&   a_lengths,
+                      std::vector<std::size_t> const&   a_strides,
+                      const int32_t                     modeA[],
+                      std::vector<std::size_t> const&   b_lengths,
+                      std::vector<std::size_t> const&   b_strides,
+                      const int32_t                     modeB[],
+                      const hipDataType                 typeScalar) override
+        {
+            using Base   = PermutationSolution;
+            using Traits = MetaTraits<DeviceOp>;
+
+            // Clear out the previous arguments
+            resetArgs();
+
+            // Promote to derived class for necessary functions such as
+            // MakeArgumentPointer and MakeInvokerPointer.
+            auto* deviceOp = dynamic_cast<DeviceOp*>(Base::mDeviceOp.get());
```

Comment:
generally speaking, hiptensor does not throw exceptions. 

If we have to use dynamic_cast here , better to use try..catch to capture the exception thrown by it.  And return an error code.

Reply:
dynamic_cast between only references throw exceptions(and supposedly not between pointers). If the casting between pointers is a failure, it just returns a nullptr. Added condition for it.